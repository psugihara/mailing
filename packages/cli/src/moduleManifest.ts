/*
 * This file is only directly used in development and tests.
 * In production is it overwritten by the build process.
 */

import sendMailFromEmails from "./emails";
import Welcome from "./emails/Welcome";
import * as WelcomePreview from "./emails/previews/Welcome";

const sendMail = sendMailFromEmails;
const previews = {
  Welcome: WelcomePreview,
};
const templates = {
  Welcome,
};
const config = {};

const manifest = { sendMail, config, templates, previews };
export { sendMail, config, templates, previews };
export default manifest;
