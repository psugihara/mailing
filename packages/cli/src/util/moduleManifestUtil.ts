import { JSXElementConstructor, ReactElement } from "react";
import { previews, config } from "../moduleManifest";

export function previewTree(): [string, string[]][] {
  return Object.keys(previews).map((previewName: string) => {
    const m = previews[previewName as keyof typeof previews];
    return [previewName, Object.keys(m)];
  });
}

export function getPreviewModule(name: string) {
  return previews[name as keyof typeof previews];
}

export function getPreviewComponent(
  name: string,
  functionName: string
): ReactElement<any, string | JSXElementConstructor<any>> | undefined {
  const previewModule:
    | {
        [key: string]: () => ReactElement | undefined;
      }
    | undefined = previews[name as keyof typeof previews] as any;
  return previewModule?.[functionName]?.();
}

export function getConfig(): MailingConfig {
  return config;
}
