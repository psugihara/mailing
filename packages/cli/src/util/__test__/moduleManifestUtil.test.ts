jest.mock("../../moduleManifest");

import moduleManifest from "../../moduleManifest";
import { previewTree } from "../moduleManifestUtil";

describe("moduleManifestUtil", () => {
  const ogPreviews = moduleManifest.previews;

  afterEach(() => {
    moduleManifest.previews = ogPreviews;
  });

  it("the original module manifest has a tree of preview components", () => {
    const previews = previewTree();
    expect(previews).toEqual([
      ["AccountCreated", ["accountCreated"]],
      ["NewSignIn", ["newSignIn"]],
      [
        "Reservation",
        ["reservationWithError", "reservationConfirmed", "reservationChanged"],
      ],
      ["ResetPassword", ["resetPassword"]],
    ]);
  });

  it("does not error on empty preview files", () => {
    moduleManifest.previews = { AccountCreated: {} } as any;
    expect(previewTree()[0]).toEqual(["AccountCreated", []]);

    moduleManifest.previews = { AccountCreated: false } as any;
    expect(previewTree()[0]).toEqual(["AccountCreated", []]);

    moduleManifest.previews = { AccountCreated: { default: jest.fn() } } as any;
    expect(previewTree()[0]).toEqual(["AccountCreated", []]);

    moduleManifest.previews = { AccountCreated: { __esModule: true } } as any;
    expect(previewTree()[0]).toEqual(["AccountCreated", []]);
  });
});
