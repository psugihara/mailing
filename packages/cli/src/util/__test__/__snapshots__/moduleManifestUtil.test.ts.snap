// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moduleManifestUtil the original module manifest has a tree of preview components 1`] = `
[
  [
    "AccountCreated",
    [
      "accountCreated",
      "accountCreatedJohn",
    ],
  ],
  [
    "NewSignIn",
    [
      "newSignIn",
    ],
  ],
  [
    "Reservation",
    [
      "reservationWithError",
      "reservationConfirmed",
      "reservationChanged",
    ],
  ],
  [
    "ResetPassword",
    [
      "resetPassword",
    ],
  ],
]
`;
